{% extends "layout.html" %}
{% block content %}
	<div class="content-section">
		<form method="POST" action="" enctype="multipart/form-data">
			{{ form.hidden_tag() }}
			<fieldset class="form-group">
				<legend class="border-bottom mb-4">Inputs for MASON</legend>
				<div class="form-group">
					{{ form.custom_id.label(class="form-control-label", style="font-weight: bold;") }}
					<p>Please select a unique custom ID for your result storage</p>
					{% if form.custom_id.errors %}
						{{ form.custom_id(class="form-control form-control-lg is-invalid") }}
						<div class="invalid-feedback">
							{% for error in form.custom_id.errors %}	
								<span>{{ error }}</span>
							{% endfor %}						
						</div>
					{% else %}
						{{ form.custom_id(class="form-control form-control-lg") }}
					{% endif %}
				</div>
				<br>
				<div class="form-group">
					{{ form.presets.label(class="form-control-label", style="font-weight: bold;") }}
					<p>Select one of the preset genomes or use "Own files" to upload your own genome.</p>
					<p> For the preset genomes of <i> Salmonella </i> (substr. SL1344) (<a href="../static/data/presets/s_typhi.fa">FASTA</a>,
						<a href="../static/data/presets/s_typhi.gff">GFF</a>) and <i> E. coli </i> (K12) (<a href="../static/data/presets/e_coli_K12.fna">FASTA</a>,
						<a href="../static/data/presets/e_coli_K12.gff">GFF</a>)we
					provide a list of essential genes. For <i> Clostridium difficile </i> 630 (<a href="../static/data/presets/c_diffi630.fasta">FASTA</a>,
						<a href="../static/data/presets/c_diffi630.gff3">GFF</a>) and
					<i> Fusobacterium nucleatum </i> (ATCC 23726) (<a href="../static/data/presets/Fuso.fasta">FASTA</a>,
						<a href="../static/data/presets/Fuso.gff3">GFF</a>) we provide the newest genome files with annotations.
						Other genomes (FASTA) and their annotations (GFF) can be downloaded from the <a href="https://www.ncbi.nlm.nih.gov/">NCBI </a>website
					and uploaded using the "Own files" option. For more details on  how to download custom genomes, have a look at our
						<a href="{{ url_for('help') }}">help page</a></p>
					{% if form.presets.errors %}
						{{ form.presets(class="form-control form-control-lg is-invalid") }}
						{% for error in form.presets.errors %}
							<span class="text-danger">{{ error }}</span><br>
						{% endfor %}
					{% else %}
						{{ form.presets(class="form-control form-control-lg") }}
					{% endif %}
				</div>
				<br>
				<div class="form-group" id="essgenes">
					{{ form.essential.label(class="form-control-label", style="font-weight: bold;") }}
					<p>Hold down the Ctrl (windows) or Command (Mac) button to select multiple genes. Please select
						no more than 10 genes at once to keep the running time low</p>
					{% if form.essential.errors %}
						{{ form.essential(class="form-control form-control-lg is-invalid") }}
						{% for error in form.essential.errors %}
							<span class="text-danger">{{ error }}</span><br>
						{% endfor %}
					{% else %}
						{{ form.essential(class="form-control form-control-lg") }}
					{% endif %}
				</div>
				<script>
					let select_presets = document.getElementById('presets');
					document.getElementById('essgenes').style.display = "none";

					let essential_choice = document.getElementById('essential');
					let essential_dict = {{ essential | tojson }};

					if (select_presets.value in essential_dict) {
						let new_choices = essential_dict[select_presets.value];
						let optionHTML = '';
						for (let item of new_choices) {
							optionHTML += '<option value="' + item.value +'">'+ item.display + '</option>';
						}
						essential_choice.innerHTML = optionHTML;
					}
					select_presets.onchange = function() {
						value = select_presets.value;
						first = document.getElementById('div_genome');
						second = document.getElementById('div_gff');
						egenes = document.getElementById('essgenes');
						if (value == 'upload') {
							egenes.style.display = 'none';
							first.style.display = 'block';
							second.style.display = 'block';
							essential_choice.innerHTML = '';
						} else {
							if (select_presets.value in essential_dict) {
								egenes.style.display = 'block';
								var new_choices = essential_dict[select_presets.value];
								essential_choice.innerHTML = '';
								let optionHTML = '';
								for (let item of new_choices) {
									optionHTML += '<option value="' + item.value +'">'+ item.display + '</option>';
								}
								essential_choice.innerHTML = optionHTML;
							}
							else {
								essential_choice.innerHTML = '';
							}
							first.style.display = 'none';
							second.style.display = 'none';
						}
					}
				</script>
				<br>
				<div class="form-group row" id="div_genome">
					<p class="col-md-12"> Below, custom FASTA and GFF files of a bacterium of interest can be uploaded (can be downloaded from
							<a href="https://www.ncbi.nlm.nih.gov/genome/">NCBI</a>): </p>
					<div class="col-md-6" id="div_genome" >
						{{ form.genome.label(class="form-control-label") }}
						{% if form.genome.errors %}
							{{ form.genome(class="form-control form-control-lg is-invalid") }}
							{% for error in form.genome.errors %}
								<span class="text-danger">{{ error }}</span><br>
							{% endfor %}
						{% else %}
							{{ form.genome(class="form-control form-control-lg") }}
						{% endif %}
					</div>
					<div class="col-md-6" id="div_gff" >
						{{ form.gff.label(class="form-control-label") }}
						{% if form.gff.errors %}
							{{ form.gff(class="form-control form-control-lg is-invalid") }}
							{% for error in form.gff.errors %}
								<span class="text-danger">{{ error }}</span><br>
							{% endfor %}
						{% else %}
							{{ form.gff(class="form-control form-control-lg") }}
						{% endif %}
					</div>
				</div>
				<br>
				<div class="form-group" id="div_lt" >
					{{ form.genes.label(class="form-control-label", style="font-weight: bold;") }}
					<p>For example: SL1344_1133, SL1344_P3_0012 for <i> Salmonella</i> SL1344. Locus tags of genes can be found in the GFF file in
						the 9th column, following "locus_tag=", see <a href="./help">help page</a>. Please select
						no more than 10 genes at once to keep the running time low </p>
					{% if form.genes.errors %}
						{{ form.genes(class="form-control form-control-lg is-invalid") }}
						{% for error in form.genes.errors %}	
							<span class="text-danger">{{ error }}</span><br>
						{% endfor %}
					{% else %}
						{{ form.genes(class="form-control form-control-lg") }}
					{% endif %}
				</div>
				<br>
				<div class="form-group">
					{{ form.len_PNA.label(class="form-control-label", style="font-weight: bold;") }}
					<p>Length of bacterial ASOs is usually chosen to be between 9-12 nucleobases to enable entry into the
						cell. Length needs to between 7 and 16 bases</p>
					{% if form.len_PNA.errors %}
						{{ form.len_PNA(class="form-control form-control-lg is-invalid") }}
						{% for error in form.len_PNA.errors %}
							<span class="text-danger">{{ error }}</span><br>
						{% endfor %}
					{% else %}
						{{ form.len_PNA(class="form-control form-control-lg") }}
					{% endif %}
				</div>
				<br>
				<div class="form-group">
					{{ form.mismatches.label(class="form-control-label", style="font-weight: bold;") }}
					<p>Number of maximally alowed mismatches for the off-target prediction algorithm.
					For 10-mer PNAs usually three or less mismatches should be chosen, as more mismatches prevent
					PNA binding</p>
					{% if form.mismatches.errors %}
						{{ form.mismatches(class="form-control form-control-lg is-invalid") }}
						{% for error in form.mismatches.errors %}
							<span class="text-danger">{{ error }}</span><br>
						{% endfor %}
					{% else %}
						{{ form.mismatches(class="form-control form-control-lg") }}
					{% endif %}
				</div>
				<br>
				<div class="form-group">
					{{ form.bases_before.label(class="form-control-label", style="font-weight: bold;") }}
					<p>If not specified, all ASOs overlapping the start codon are designed.
					If a SD sequence exists, it can be useful to design more ASOs lying upstream (5') the CDS to inhibit
					translation </p>
					{% if form.bases_before.errors %}
						{{ form.bases_before(class="form-control form-control-lg is-invalid") }}
						{% for error in form.bases_before.errors %}
							<span class="text-danger">{{ error }}</span><br>
						{% endfor %}
					{% else %}
						{{ form.bases_before(class="form-control form-control-lg") }}
					{% endif %}
				</div>
				<br>
				<div class="form-group">
					<p><b> Screen other genomes for off-targets (optional) </b></p>

					<p> Select to screen for off-targets in the <a href="https://www.ncbi.nlm.nih.gov/genome/guide/human/">human genome</a>
						(<a href="../static/data/presets/GRCh38_latest_rna.fna">FASTA</a>) or
						the <a href="https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA28331">human microbiome</a>
						(<a href="../static/data/presets/start_regions_HMP.fasta">FASTA</a>). For the human genome,
						the whole transcriptome is screened, whereas for the human microbiome only translation
						initiation sites are considered. Note, that the selection of these additional screens will
						increase the running time of MASON</p>

					<input type="radio" name="add_screen" id="none" value="none" checked> None  </input><br>
					<input type="radio" name="add_screen" id="add_human" value="human"> Human genome  </input><br>
					<input type="radio" name="add_screen" id="add_microbiome" value="microbiome"> Human microbiome </input><br>
				<!--	<input type="radio" name="add_screen" id="add_custom" value="custom"> Custom genome </input><br>
					<br>
					<div class="form-group row" id="div_genome_2">
						<p class="col-md-12"> Optionally, custom FASTA and GFF files of another organism of interest can be uploaded to be
							screened for off-targets (can be downloaded from
								<a href="https://www.ncbi.nlm.nih.gov/genome/">NCBI</a>): </p>
						<div class="col-md-6" id="div_genome" >
							{{ form.genome.label(class="form-control-label") }}
							{% if form.genome.errors %}
								{{ form.genome(class="form-control form-control-lg is-invalid") }}
								{% for error in form.genome.errors %}
									<span class="text-danger">{{ error }}</span><br>
								{% endfor %}
							{% else %}
								{{ form.genome(class="form-control form-control-lg") }}
							{% endif %}
						</div>
						<div class="col-md-6" id="div_gff" >
							{{ form.gff.label(class="form-control-label") }}
							{% if form.gff.errors %}
								{{ form.gff(class="form-control form-control-lg is-invalid") }}
								{% for error in form.gff.errors %}
									<span class="text-danger">{{ error }}</span><br>
								{% endfor %}
							{% else %}
								{{ form.gff(class="form-control form-control-lg") }}
							{% endif %}
					</div>
					<script>
						document.getElementById('div_genome_2').style.display = "none";
						let select_none = document.getElementById('none');
						let select_human_genome = document.getElementById('add_human');
						let select_microbome = document.getElementById('add_microbiome');
						let select_add_genome = document.getElementById('add_custom');
						select_add_genome.onclick = function() {
							document.getElementById('div_genome_2').style.display = "block";
						}
						select_human_genome.onclick = function() {
							document.getElementById('div_genome_2').style.display = "none";
						}
						select_microbome.onclick = function() {
							document.getElementById('div_genome_2').style.display = "none";
						}
						select_none.onclick = function() {
							document.getElementById('div_genome_2').style.display = "none";
						}

					</script>
					-->
				</div>
				</div>
			</fieldset>
			<div class="form-group">
				{{ form.submit(class="btn btn-outline-info") }}
			</div>
		</form>
	</div>
{% endblock %}